src/fs/cluster_allocator.cc

```diff
@@ -21,34 +21,71 @@
 
 #include "fs/cluster.hh"
 #include "fs/cluster_allocator.hh"
+#include "seastar/core/future.hh"
 
 #include <cassert>
 #include <optional>
 
-namespace seastar {
+namespace seastar::fs {
 
-namespace fs {
+cluster_allocator::cluster_allocator() : _cluster_sem(0) {}
 
-cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters)
-        : _allocated_clusters(std::move(allocated_clusters)), _free_clusters(std::move(free_clusters)) {}
+void cluster_allocator::init(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters) {
+    _allocated_clusters = std::move(allocated_clusters);
+    _free_clusters = std::move(free_clusters);
+    assert(_cluster_sem.waiters() == 0);
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
+}
 
-std::optional<cluster_id_t> cluster_allocator::alloc() {
-    if (_free_clusters.empty()) {
-        return std::nullopt;
-    }
+cluster_id_t cluster_allocator::do_alloc() {
+    assert(not _free_clusters.empty());
 
     cluster_id_t cluster_id = _free_clusters.front();
     _free_clusters.pop_front();
     _allocated_clusters.insert(cluster_id);
+
     return cluster_id;
 }
 
-void cluster_allocator::free(cluster_id_t cluster_id) {
+cluster_id_t cluster_allocator::do_free(cluster_id_t cluster_id) {
     assert(_allocated_clusters.count(cluster_id) == 1);
     _free_clusters.emplace_back(cluster_id);
     _allocated_clusters.erase(cluster_id);
+    return cluster_id;
 }
 
+std::optional<cluster_id_t> cluster_allocator::alloc() {
+    if (_free_clusters.empty()) {
+        return std::nullopt;
+    }
+
+    assert(_cluster_sem.available_units() > 0);
+    _cluster_sem.consume(1);
+
+    return do_alloc();
+}
+
+future<std::vector<cluster_id_t>> cluster_allocator::alloc_wait(size_t count) {
+    return _cluster_sem.wait(count).then([this, count] {
+        std::vector<cluster_id_t> cluster_ids;
+        for (size_t i = 0; i < count; ++i) {
+            cluster_ids.emplace_back(do_alloc());
```

Comment:
If this fails we need to rollback previous allocations to provide strong exception guarantee, as it is cheap here